In [1]:
import sys
sys.path.append("..")

from pathlib import Path
import plotly.express as px
from utils.transcription_utils import get_transcription_info,get_stats,get_total_stats,get_data_loss
import csv
import pandas as pd

In [2]:
save_path=Path("/media/dataset-harddisk/munikumar/utils/audio_utils/data/gramvaani/nuktas")

In [3]:
dataset_dir=Path("/media/dataset-harddisk/munikumar/hindi_dataset/gramvaani/gramvaani_raw")

In [4]:
utt_labels = [
    "/media/dataset-harddisk/munikumar/hindi_dataset/gramvaani/gramvaani_raw/GV_Train_100h/utt2labels",
    "/media/dataset-harddisk/munikumar/hindi_dataset/gramvaani/gramvaani_raw/GV_Dev_5h/utt2labels",
    "/media/dataset-harddisk/munikumar/hindi_dataset/gramvaani/gramvaani_raw/GV_Eval_3h/utt2labels",
]

In [5]:
def load_transcription(transcription_path):
    transcriptions={}
    with open(transcription_path, "r", encoding="utf-8") as fp:
        for transcription_data in fp.readlines():
            audio_file = transcription_data.split(" ")[0]+".mp3"
            transcription = " ".join(transcription_data.split(" ")[1:-1])
            transcriptions[audio_file] = transcription
    return transcriptions

In [6]:
# dev_transcription_data=load_transcription(dataset_dir/"GV_Dev_5h"/"text")
# get_transcription_info(dev_transcription_data,dataset_dir/"GV_Dev_5h"/"Audio",save_path,"dev")
get_stats(save_path/"dev_non_hindi_transcripts.json")

**************************************************
Loss of audio data without transcription normalization:
**************************************************
Total duration : 5:4:54
Loss duration : 0:0:20
Remaining duration : 5:4:34
Drop by percentage: 0.11

Total audio files : 1885
Files that would be drop: 2
Remaining audio files : 1883
Drop by percentage: 0.11
**************************************************
Non hindi tokes : ['.', 'ऍ']


### Dev Dataset

**************************************************
Loss of audio data without transcription normalization:
**************************************************
Total duration : 5:4:54  
Loss duration : 2:4:48  
Remaining duration : 3:0:6  
Drop by percentage: 40.93  

Total audio files : 1885  
Files that would be drop: 724  
Remaining audio files : 1161  
Drop by percentage: 38.41  
**************************************************
Non hindi tokes : ['.', 'ऍ', 'ऱ', 'क़', 'ख़', 'ग़', 'ज़', 'ड़', 'ढ़', 'फ़']


**************************************************
Loss of audio data after Nuktas ['ऱ', 'क़', 'ख़', 'ग़', 'ज़', 'ड़', 'ढ़', 'फ़'] Decomposition:
**************************************************
Total duration : 5:4:54  
Loss duration : 0:0:20  
Remaining duration : 5:4:34  
Drop by percentage: 0.11  

Total audio files : 1885  
Files that would be drop: 2  
Remaining audio files : 1883  
Drop by percentage: 0.11  
**************************************************
Non hindi tokes : ['.', 'ऍ'] 

In [7]:
# test_transcription_data=load_transcription(dataset_dir/"GV_Eval_3h"/"text")
# get_transcription_info(test_transcription_data,dataset_dir/"GV_Eval_3h"/"Audio",save_path,"test")
get_stats(save_path/"test_non_hindi_transcripts.json")

**************************************************
Loss of audio data without transcription normalization:
**************************************************
Total duration : 2:49:47
Loss duration : 0:0:0
Remaining duration : 2:49:47
Drop by percentage: 0.00

Total audio files : 1032
Files that would be drop: 0
Remaining audio files : 1032
Drop by percentage: 0.00
**************************************************
Non hindi tokes : []


### Test dataset

**************************************************
Loss of audio data without transcription normalization:
**************************************************
Total duration : 2:49:47  
Loss duration : 1:18:21  
Remaining duration : 1:31:25  
Drop by percentage: 46.15  

Total audio files : 1032  
Files that would be drop: 450  
Remaining audio files : 582  
Drop by percentage: 43.60  
**************************************************
Non hindi tokes : ['क़', 'ख़', 'ग़', 'ज़', 'ड़', 'ढ़', 'फ़']


**************************************************
Loss of audio data After Nuktas decomposition:
**************************************************
Total duration : 2:49:47  
Loss duration : 0:0:0  
Remaining duration : 2:49:47  
Drop by percentage: 0.00  

Total audio files : 1032  
Files that would be drop: 0  
Remaining audio files : 1032  
Drop by percentage: 0.00  

In [8]:
# train_transcription_data=load_transcription(dataset_dir/"GV_Train_100h"/"text")
# get_transcription_info(train_transcription_data,dataset_dir/"GV_Train_100h"/"Audio",save_path,"train")
get_stats(save_path/"train_non_hindi_transcripts.json")

**************************************************
Loss of audio data without transcription normalization:
**************************************************
Total duration : 102:10:50
Loss duration : 4:33:5
Remaining duration : 97:37:45
Drop by percentage: 4.45

Total audio files : 37152
Files that would be drop: 1470
Remaining audio files : 35682
Drop by percentage: 3.96
**************************************************
Non hindi tokes : ["'", ',', '-', '.', ':', 'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M', 'N', 'O', 'P', 'R', 'S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z', '[', '\\', ']', 'a', 'b', 'd', 'e', 'h', 'i', 'k', 'l', 'm', 'n', 'p', 'r', 's', 't', 'y', 'ऍ', '।', '\u200c']


### Train Dataset
**************************************************
**Loss of audio data without transcription normalization:**
**************************************************
Total duration : 102:10:50  
Loss duration : 38:54:3  
Remaining duration : 63:16:47  
Drop by percentage: 38.07  

Total audio files : 37152  
Files that would be drop: 13180  
Remaining audio files : 23972  
Drop by percentage: 35.48  
**************************************************
Non hindi tokes : ["'", ',', '-', '.', ':', 'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M', 'N', 'O', 'P', 'R', 'S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z', '[', '\\', ']', 'a', 'b', 'd', 'e', 'h', 'i', 'k', 'l', 'm', 'n', 'p', 'r', 's', 't', 'y', 'ऍ', 'ऱ', 'क़', 'ख़', 'ग़', 'ज़', 'ड़', 'ढ़', 'फ़', '।', '\u200c']

**************************************************
**************************************************
**Loss of audio data After Decomposing Nuktas:**
**************************************************
Total duration : 102:10:50  
Loss duration : 4:33:5  
Remaining duration : 97:37:45  
Drop by percentage: 4.45  

Total audio files : 37152  
Files that would be drop: 1470  
Remaining audio files : 35682  
Drop by percentage: 3.96  
**************************************************
Non hindi tokes : ["'", ',', '-', '.', ':', 'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M', 'N', 'O', 'P', 'R', 'S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z', '[', '\\', ']', 'a', 'b', 'd', 'e', 'h', 'i', 'k', 'l', 'm', 'n', 'p', 'r', 's', 't', 'y', 'ऍ', '।', '\u200c']
 

In [9]:
get_data_loss(396331,10680,40069,1017)

Total duration : 110:5:31
Loss duration : 2:58:0
Remaining duration : 107:7:31
Drop by percentage: 2.69

Total audio files : 40069
Files that would be drop: 1017
Remaining audio files : 39052
Drop by percentage: 2.54


In [10]:
get_total_stats(save_path)

**************************************************
Loss of audio data for total dataset:
**************************************************
Total duration : 110:5:32
Loss duration : 4:33:25
Remaining duration : 105:32:7
Drop by percentage: 4.14

Total audio files : 40069
Files that would be drop: 1472
Remaining audio files : 38597
Drop by percentage: 3.67


### Total Dataset
**************************************************
**Loss of audio data for total dataset:**
**************************************************
Total duration : 110:5:32  
Loss duration : 42:17:13  
Remaining duration : 67:48:19  
Drop by percentage: 38.41  

Total audio files : 40069  
Files that would be drop: 14354  
Remaining audio files : 25715  
Drop by percentage: 35.82  
 

**************************************************
**Loss of audio data for total dataset After decomposing nuktas:**
**************************************************
Total duration : 110:5:32  
Loss duration : 4:33:25  
Remaining duration : 105:32:7  
Drop by percentage: 4.14  

Total audio files : 40069  
Files that would be drop: 1472  
Remaining audio files : 38597  
Drop by percentage: 3.67   

## Visualizing the metadata

In [15]:
all_data=[]
utt_labels=list(Path("/media/dataset-harddisk/munikumar/datasets/hindi_dataset/gramvaani/gramvaani_raw").rglob("*utt2labels"))

In [17]:
for utt_label in utt_labels:
    with open(utt_label,'r')as fp:
        reader=csv.reader(fp,delimiter='\t')
        all_data.append(list(reader))

In [20]:
data=all_data[0]+all_data[1][1:]+all_data[2][1:]
column=data[0]
df=pd.DataFrame(data[1:],columns=column)

In [22]:
df.head()

Uttids  Accent    Age Gender Background Sentiment District  State  \
0  01-00003-01      NA  Adult   Male         NA        NA       NA     NA   
1  01-00005-01  Bihari  Adult   Male         NA   Serious       NA  Bihar   
2  01-00008-02  Bihari  Adult   Male         NA   Serious       NA  Bihar   
3  01-00009-01  Bihari  Adult   Male         NA   Serious       NA  Bihar   
4  01-00014-02  Bihari  Adult   Male         NA   Serious       NA  Bihar   

       Other  
0         NA  
1  inaudible  
2  inaudible  
3  inaudible  
4  inaudible

### Gender Distribution

In [23]:
gender_counts=df["Gender"].value_counts()
gender_counts

Gender
Male              26212
Female            11288
NA                 2530
Gender_unknown       39
Name: count, dtype: int64

In [32]:
fig = px.pie(values=gender_counts, names=gender_counts.index, title='Gender Distribution')
fig.show()

### Accents Distribution

In [33]:
accent_counts=df["Accent"].value_counts()
accent_counts

Accent
Accent_unknown    19526
Bihari            12788
NA                 6880
Bhojpuri            822
Normal_hindi         19
Jharkhandi           16
Bengali               8
Marathi               6
Lucknowi              4
Name: count, dtype: int64

In [34]:
# accent_counts.drop("Accent_unknown",inplace=True)
# accent_counts.drop("NA",inplace=True)

In [35]:
px.pie(values=accent_counts, names=accent_counts.index, title='Accent Distribution').show()


### Age Distribution

In [36]:
age_counts=df["Age"].value_counts()
age_counts

Age
Adult          35393
NA              4163
Age_unknown      296
Child            217
Name: count, dtype: int64

In [37]:
px.pie(values=age_counts, names=age_counts.index, title='age Distribution').show()


### Noise Distribution

In [38]:
background_counts=df["Background"].value_counts()
background_counts

Background
NA             36027
Music           3964
Noise             66
No_speaking       12
Name: count, dtype: int64

In [39]:
px.pie(values=background_counts, names=background_counts.index, title='background Distribution').show()


### Sentiment Distribution

In [40]:
sentiment_counts=df["Sentiment"].value_counts()
sentiment_counts

Sentiment
NA         32977
Serious     6544
Sad          360
Happy        171
Angry         16
Tense          1
Name: count, dtype: int64

In [41]:
px.pie(values=sentiment_counts, names=sentiment_counts.index, title='sentiment Distribution').show()


### State Distribution

In [42]:
state_counts=df["State"].value_counts()
state_counts

State
NA                27059
Bihar             12876
Jharkhand            61
Madhya_pradesh       45
Maharashtra          11
West_bengal           8
Uttar_pradesh         4
Rajasthan             3
New_delhi             2
Name: count, dtype: int64

In [43]:
px.pie(values=state_counts, names=state_counts.index, title='state Distribution').show()
